### PR TITLE
set scalac -Xsource:2.13 -Ypartial-unification globally

### PR DESCRIPTION
**Pull request by S11001001**
These options are commonly added to make things nicer, and the behavior is anyway closer to what Scala 2.13 is like.

Has come up in discussion variously with @stefanobaghino-da , @remyhaemmerle-da @rohanjr .

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: bazel_tools/scala.bzl [hunk 1]
```diff
@@ -38,6 +38,8 @@ common_scalacopts = [
     # warn if using deprecated stuff
     "-deprecation",
     "-Xfuture",
+    "-Xsource:2.13",  # fix typechecker bugs
+    "-Ypartial-unification",  # behave more like Scala 2.13
```
Comment (S11001001) [inline bazel_tools/scala.bzl:42]: These comments are not backwards.

Comment (stefanobaghino-da) [reply bazel_tools/scala.bzl:42]: If you had to comment your comment, probably it's a sign the comments by themselves are not sufficient. Adding more details could make it easier on the readers without having to know about (or remember) this PR. If it's too verbose, a simple pointer to a URL with an understandable explanation is probably good enough.

**Comment by S11001001** [comment]
> How come `scalaz` became a direct dependency of some many packages? It doesn't look like you added any use in the PR.

@stefanobaghino-da For errors like this.

```
daml-lf/interface/src/test/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReaderSpec.scala:39: error: Symbol 'type scalaz.Order' is missing from the classpath.
This symbol is required by 'method com.daml.lf.data.IdString.Party$u0020order$u0020instance'.
Make sure that type Order is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
A full rebuild may help if 'IdString.class' was compiled against an incompatible version of scalaz.
            ImmArray[Ref.Name]("call", "put").toSeq,
                                       ^
```

It is already the case that various bazel files in the repo reference libraries "directly" for this reason. Why is it triggered only when the options are added? My best guess is that when scala/bug#10283 is fixed (as indeed is the primary benefit of `-Xsource:2.13`), this method can be included in some implicit searches (albeit rejected).

On the other hand, I have no good theory whatsoever why the `Conversions._` triggered unused import warning but did not without the flags.

> I'm happy to make this step, but I believe this PR should also remove all explicit usages of `-Xsource:2.13` and `-Ypartial-unification` as well.

Agreed; will not mark this PR as ready for review until that is done.

File: bazel_tools/scala.bzl (hunk 1 above)
Comment (S11001001) [reply bazel_tools/scala.bzl:42]: I looked when you were asking about this before, but there was nothing coherent I could link to. I'll just put a description in the comments.

Comment (S11001001) [reply bazel_tools/scala.bzl]: e5cb754f850371244684cecb16121839495dc0b0

**Comment by S11001001** [comment]
@rohanjr As it happens, those particular changes would work without the new scalac options, too. It's just not worth dealing with `Future.traverse`'s signature problems, especially four times in the same file.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala [hunk 2]
```diff
@@ -225,14 +218,8 @@ object Runner {
...
     for {
-      // Apparently Future.traverse is too stupid to traverse an
-      // Option so you first have to convert to a list and then convert back.
-      defClient <- Future
-        .traverse(participantParams.default_participant.toList)(client(_))
-        .map(_.headOption)
-      otherClients <- Future
-        .traverse(participantParams.participants)({ case (k, v) => client(v).map(c => (k, c)) })
-        .map(_.toMap)
+      defClient <- participantParams.default_participant.traverse(client(_))
```
Comment (leo-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala:221]: what happened with `headOption` it is not required any longer.... `participantParams.default_participant` is `Option[A]`, right?

Comment (S11001001) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala:221]: Indeed.

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/UsedTypeParams.scala [hunk 3]
```diff
@@ -23,7 +23,8 @@ object UsedTypeParams {
 
   private def foldMapGenTypes[Z: Monoid](typeDecl: DataType)(f: iface.Type => Z): Z = {
     val notAGT = (s: String) => mzero[Z]
-    typeDecl.foldMap(_.bifoldMap(f)(_.bifoldMap(_ foldMap (_.bifoldMap(notAGT)(f)))(f)))
+    (typeDecl: ScopedDataType[iface.DataType[iface.Type, VariantField]])
```
[the remark this answers is not in the document]
Comment (S11001001) [reply language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/UsedTypeParams.scala:26]: Aliases have never been 100% well-cooked in Scala. 🤷‍♂️

**Pull request merged**